packages/flutter/lib/src/material/text_button.dart

```diff
@@ -224,6 +224,9 @@ class TextButton extends ButtonStyleButton {
     final MaterialStateProperty<MouseCursor>? mouseCursor = (enabledMouseCursor == null && disabledMouseCursor == null)
       ? null
       : _TextButtonDefaultMouseCursor(enabledMouseCursor!, disabledMouseCursor!);
+    final MaterialStateProperty<BorderSide?>? _side = (side is MaterialStateProperty
+      ? side
+      : ButtonStyleButton.allOrNull<BorderSide>(side)) as MaterialStateProperty<BorderSide?>?;
```

Comment:
This code works, but I have a strong suspicion it isn't quite _right_. I was finding that in the following code sample, providing MaterialStateProperty instances were not dynamically resolved:

```dart
TextButtonTheme(
  data: TextButton.styleFrom(
    // Did not work - no dynamic state resolution
    side: MaterialStateBorderSide.resolveWith(getBorderSide),
  )
  child: TextButon(...)
)
```

This seems to be because `TextButton.styleFrom()` only accepts primitive instances, not `MaterialStateProperty<XYZ>` instances, and MaterialStateProperties spell out in their docstrings that "if they are used as regular XYZs, the default value will be used instead."

And so I made this change to `TextButton.styleFrom()` to ensure that MaterialStateResolvers from ancestor themes were dynamically resolved. However, this feels like either a gotcha, or a section of the puzzle I didn't fully put together.

Reply:
Oh right, I almost forgot to say - another reason I suspect this code is not right is that it uses `is` and `as` - both frowned upon by the style guide.

Reply:
No changes are needed here, because ButtonStyle supports a `MaterialStateProperty<BorderSide>` already.  The TextButton's styleOf method is a convenience for the case where a `MaterialStateProperty<BorderSide>` isn't required, because the same border is to be used for all states. The MaterialStateBorderSide class, like MaterialStateColor, is just for the fairly rare case of an existing API that was originally defined in terms of BorderSide and now must also (optionally) deal with `MaterialStateProperty<BorderSide>` values. Currently there appears to be just one example of that, in Chip and ChipTheme.

Note also: there's no need to find existing widgets that could also support MaterialBorderSide. This class is just a pressure relief valve :-).

Reply:
Thanks for this clarification. I've reverted the `TextButton` form of this and added a test for `ChipTheme`'s usage of `MaterialStateBorderSide.resolveWith()`.